animeshon/knowledge/v1alpha1/knowledge.proto

```diff
@@ -18,66 +19,123 @@ service Knowledge {
     option (google.api.http) = {
-      get: "/v1alpha1/contributions"
+      get: "/v1alpha1/{parent=users/*}/contributions"
     };
   }
 
   rpc GetContribution(GetContributionRequest) returns (Contribution) {
     option (google.api.http) = {
-      get: "/v1alpha1/{name=contributions/*}"
+      get: "/v1alpha1/{name=users/*/contributions/*}"
+    };
+  }
+
+  rpc GetContributionChanges(GetContributionChangesRequest) returns (ContributionChanges) {
+    option (google.api.http) = {
+    get: "/v1alpha1/{name=users/*/contributions/*/changes}"
     };
   }
 
   rpc ApproveContribution(ApproveContributionRequest) returns (Contribution) {
     option (google.api.http) = {
-      post: "/v1alpha1/{name=contributions/*}:approve"
+      post: "/v1alpha1/{name=users/*/contributions/*}:approve"
+      body: "*"
+    };
+  }
+
+  rpc ReviewContribution(ReviewContributionRequest) returns (Contribution) {
+    option (google.api.http) = {
+      patch: "/v1alpha1/{name=users/*/contributions/*}"
       body: "*"
     };
   }
 
   rpc RejectContribution(RejectContributionRequest) returns (Contribution) {
     option (google.api.http) = {
-      post: "/v1alpha1/{name=contributions/*}:reject"
+      post: "/v1alpha1/{name=users/*/contributions/*}:reject"
       body: "*"
     };
   }
 }
 
-message Contribution {
-  option (google.api.resource) = {
-    type: "knowledge.animeapis.com/Contribution"
-    pattern: "contributions/{contribution}"
-  };
+message ContributionChanges {
+  repeated EntryEntity additions = 1;
 
+  repeated EntryEntity deletions = 2;
+}
+
+message Contribution {
+  enum State {
+    PENDING = 0;
+    APPROVED = 1;
+    REJECTED = 2;
+    DRAFT = 3;
+  }
   // The resource name of the contribution.
-  string name = 1;
+  string name = 1 [(google.api.field_behavior) = REQUIRED];
+
+  // The parent resource where this contribution will be created.
+  string parent = 2 [(google.api.field_behavior) = REQUIRED];
+
+  string reviewer = 3;
+
+  State state = 4;
+
+  string display_name = 5;
+
+  int32 revision = 6;
+
+  // When the album was created.
+  google.protobuf.Timestamp create_time = 7;
+
+  // When the album was updated.
+  google.protobuf.Timestamp update_time = 8;
+
+  // When the album was reviewd.
+  google.protobuf.Timestamp review_time = 9;
+}
+
+message ReviewContributionRequest {
+  // The parent resource where this contribution will be created.
+  string parent = 1 [(google.api.field_behavior) = REQUIRED];
+
+  string name = 2;
+
+  repeated EntryEntity additions = 3;
+
+  repeated EntryEntity deletions = 4;
 }
 
 message CreateContributionRequest {
-  // The resource name.
-  string name = 1;
+  // The parent resource where this contribution will be created.
+  string parent = 1 [(google.api.field_behavior) = REQUIRED];
+
+  string display_name = 2;
 
-  // The resurce payload.
-  bytes payload = 2;
+  repeated EntryEntity additions = 3;
```

Comment:
Accept the message `ContributionChanges` as `ContributionChanges changes = 3;` instead.